Add `asyncWhere`

natebosch opened the pull request:
Relatively few tests since most of the complex stuff is handled by
StreamTransformer.fromHandlers

natebosch committed:
Add `asyncWhere`

Relatively few tests since most of the complex stuff is handled by
StreamTransformer.fromHandlers

grouma reviewed (approved):
Optionally add a test with an explicit microtask.

natebosch committed:
Handle predicates that go around the event loop

- Use a counter for values waiting since more than 1 can come in before
the close can be handled
- synchronously recorde the valuesWaiting so we don't close before
noticing any values

natebosch commented:
Good catch - there were bugs when I didn't go around the event loop
